### PR TITLE
tapdb: add minimalistic test framework for testing DB migrations

### PR DESCRIPTION
**Pull request by guggero**
This is an attempt at solving https://github.com/lightninglabs/taproot-assets/issues/694.
There are many ways to solve this, and this is IMO the most easy to implement one.

So this is mainly a proposal to get the discussion started. If we feel like we need something more involved, we can discuss based on this example code.

I mainly had this PR in mind when coming up with the design for this: https://github.com/lightninglabs/taproot-assets/pull/684
There we need to have a way to insert existing data into the DB before applying the latest migration.

Here are the pros/cons of the current approach:

Pros:
 - Simple to implement, understand, use and maintain
 - Should allow us to fully test data manipulation migrations
 - No need to carry along multiple versions of Golang model code
 - No impact on production code or run time

Cons:
 - Not very easy to test schema manipulation migrations, since Golang model is always the very latest version
 - If we need to test changes to the actual schema, we need to do it with raw SQL queries
 - Need to manually create test data SQL code

Fixes https://github.com/lightninglabs/taproot-assets/issues/694

### CODE REVIEW

**Comment by bhandras** [comment]
Super cool!

**Review by ffranr — APPROVED** [review]
Migration changes make sense to me. Nice PR, very clear! I'm not sure where you're going with the last commit.

**Comment by guggero** [comment]
> e.g if we already have 10 migrations in the DB and a new PR is trying to add n.11, are we gonna test that migration against all possible states a tapd could be upgrading from?

The idea is that we have a way to test migrations that do data manipulations. Those are quite hard to test with just a unit test that looks at the state _after_ all migrations have been applied. One of those PRs is https://github.com/lightninglabs/taproot-assets/pull/684 (which was taken as the inspiring PR to influence some of the design in this PR).

**Review by jharveyb — APPROVED** [review]
Very neat! I think this would have worked for validating the latest changes to the `anchor_txid` view, and repro'ing that user issue.

Re: test data; I imagine we could generate a 'golden' DB state from one of itests and use that for testing future migrations. Feels like using this in future tests shouldn't be too much work.

Needs rebase.

File: tapdb/assets_store.go [hunk 1]
```diff
@@ -861,7 +861,7 @@ func fetchAssetsWithWitness(ctx context.Context, q ActiveAssetsStore,
 	// First, we'll fetch all the assets we know of on disk.
 	dbAssets, err := q.QueryAssets(ctx, assetFilter)
 	if err != nil {
-		return nil, nil, fmt.Errorf("unable to read db assets: %v", err)
+		return nil, nil, fmt.Errorf("unable to read db assets: %w", err)
```
Comment (Roasbeef) [inline tapdb/assets_store.go:864]: 👍

**Review by Roasbeef — APPROVED** [review]
Was a bit skeptical initially, but I think this nicely resolves issues that would pop up in a _purely_ unit test approach, namely: we wouldn't be able to programmatically create older versions of certain tables as the Go code is generated from the _latest_ version of the migration. 

It's also the case that all the generated code already has SQL fragments in them, and you can look at the older versions to extract raw SQL used for the insertion files. 

LGTM 🥕

**Pull request merged**